### PR TITLE
Added PipelineState and used it for user data nodes

### PR DESCRIPTION
**Pull request by trenouf**
This commit is one step towards two goals of the Builder interface:
    a. Make Builder the interface between the frontend and the middle-end of
       LLPC, without any on-the-side context data structures.
    b. Allow the middle-end (Builder and Patch) to work without needing any
       outside-IR persistent state, so we can add and use proposed
       -stop-before, -start-before and -run-pass options to amdllpc and test
       individual passes.
    
    The changes in this commit are:
    
    1. There is a new PipelineState object, designed to hold pipeline state
       (i.e. things that are supplied from the frontend and do not change
       through the middle-end). In this commit, it only contains the user
       data nodes.
    
    2. State from PipelineState is written into IR as metadata. (Again,
       currently only the user data nodes.)
    
    3. There is a new PipelineStateWrapper immutable pass used to query the
       pipeline state. The first time you use it, it reads the state out of
       the IR and caches it for the rest of the compile. Because it is an
       immutable pass, it is always considered preserved by other passes.
    
    4. Patch passes that access the user data nodes now use
       PipelineStateWrapper.
    
    5. There is a new Builder method to set user data nodes from the
       frontend.
    
    Change-Id: Ia7264874e4f0a9dce1927a54bb2a11df644a0ac3

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit aff4f12dcc508cedec7d8f49c64275e64727ed18
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev 7438ae21a2e8e3c782f82745249da539a19c193e</li>
<li>SPVGEN: dev 5378c7618a2a3bf60ba01a86db980930e8581146</li>
<li>XGL: dev 9490055b71ec89be68c1ed4a3a847569c0ac3cbf</li>
<li>LLVM: amd-vulkan-dev f41e1a873108a371ae5574d518c1ee6eb3814cee</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 893/893 (100.0%)</li>
<li>Failed: 0/893 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 3740/3741 (100.0%)</li>
<li><details><summary>Failed: 1/3741 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.object_management.single.graphics_pipeline
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>


</ul></details>

**Comment by JacobHeAMD** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit aff4f12dcc508cedec7d8f49c64275e64727ed18
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev 9a1f99f049f3b417242dc7c09b86c52fd7d9ceaf</li>
<li>SPVGEN: dev 64013f150c2d41965a3d36fa159b55fba52b727a</li>
<li>XGL: dev 4d91a4045078cf6884ee05007eac9eec2b881dbb</li>
<li>LLVM: amd-vulkan-dev 4305f776f1207189667a383e22661ccd1e99cd25</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 893/893 (100.0%)</li>
<li>Failed: 0/893 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 3740/3741 (100.0%)</li>
<li><details><summary>Failed: 1/3741 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.object_management.single.graphics_pipeline
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
@trenouf , please resolve conficts

**Comment by amdrexu** [comment]
Please update the commits based on internal discussion and merge those commits to one. Thank you.

**Comment by JacobHeAMD** [comment]
retest this please

**Comment by JaxLinAMD** [comment]
retest this please

File: util/llpcUtil.cpp [hunk 1]
```diff
@@ -154,4 +154,37 @@ bool CreateDirectory(
 #endif
 }
 
+// =====================================================================================================================
+// Helper macro
+#define CASE_CLASSENUM_TO_STRING(TYPE, ENUM) \
```
Comment (amdrexu) [inline util/llpcUtil.cpp]: Put this macro and the function GetResourceMappingNodeTypeName in llpcInternal.cpp
For llpcUtil.cpp, we often place simple graphics-independent functions, such as math functions.

File: patch/llpcPatchEntryPointMutate.cpp [hunk 2]
```diff
@@ -408,14 +408,27 @@ FunctionType* PatchEntryPointMutate::GenerateEntryPointType(
              // and indirect user data should not be counted in possible spilled user data.
             if (pNode->type == ResourceMappingNodeType::IndirectUserDataVaPtr)
             {
-                // Only the first shader stage needs a vertex buffer table.
-                if ((m_pContext->GetShaderStageMask() & (ShaderStageToMask(m_shaderStage) - 1)) == 0)
+                // Only the vertex shader needs a vertex buffer table.
```
Comment (amdrexu) [inline patch/llpcPatchEntryPointMutate.cpp]: This commit shouldn't be included in this pull request, right? I see you have made a separate one.

File: builder/llpcPipelineState.cpp [hunk 3]
```diff
@@ -28,3 +28,408 @@
...
+                        for (uint32_t compIdx = 0; compIdx < SamplerDescriptorSize; ++compIdx)
+                        {
+                            compValues[compIdx] = mdconst::dyn_extract<ConstantInt>(
+                                  pMetadataNode->getOperand(OperandStartIdx + SamplerDescriptorSize * elemIdx + compIdx));
+                        }
+                        descriptors.push_back(ConstantVector::get(compValues));
+                    }
+                    pNextNode->pImmutableValue = ConstantArray::get(ArrayType::get(descriptors[0]->getType(),
+                                                                                   elemCount),
+                                                                    descriptors);
```
[the remark this answers is not in the document]
Comment (trenouf) [reply builder/llpcPipelineState.cpp]: What is wrong with this alignment? descriptors is the second arg to ConstantArray::get.

[the remark this answers is not in the document]
Comment (amdrexu) [reply builder/llpcPipelineState.cpp]: Or, it is my fault. Ignore this.

File: patch/llpcPatchEntryPointMutate.cpp (hunk 2 above)
Comment (trenouf) [reply patch/llpcPatchEntryPointMutate.cpp]: Correct, but I have to include both commits in the pull request for the testing to work. Once pr89 has been pulled, I can rebase this pr so only one commit remains.

Comment (amdrexu) [reply patch/llpcPatchEntryPointMutate.cpp]: Understand. You can make comments in the pull request :)

Comment (trenouf) [reply patch/llpcPatchEntryPointMutate.cpp]: (I mean pr88.)

**Comment by JacobHeAMD** [comment]
Hi @trenouf , please help to resolve the conflicts and rebase to the latest dev. Sorry for delay.

**Comment by trenouf** [comment]
Thanks Jacob; now resolved.

No problem about delay; I guess the two main reasons were:
1. your office move; and
2. me breaking everyone's testing with my "Merge user data nodes" change. Sorry about that.

**Comment by amdvlk-admin** [comment]
### Test summary for commit de5626c687dac247981696564a95a25c531a6691
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev 9a1f99f049f3b417242dc7c09b86c52fd7d9ceaf</li>
<li>SPVGEN: dev 64013f150c2d41965a3d36fa159b55fba52b727a</li>
<li>XGL: dev 4d91a4045078cf6884ee05007eac9eec2b881dbb</li>
<li>LLVM: amd-vulkan-dev 4305f776f1207189667a383e22661ccd1e99cd25</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 900/900 (100.0%)</li>
<li>Failed: 0/900 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 3740/3741 (100.0%)</li>
<li><details><summary>Failed: 1/3741 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.object_management.single.graphics_pipeline
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 3740/3741 (100.0%)</li>
<li><details><summary>Failed: 1/3741 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.object_management.single.graphics_pipeline
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit de5626c687dac247981696564a95a25c531a6691
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev 9a1f99f049f3b417242dc7c09b86c52fd7d9ceaf</li>
<li>SPVGEN: dev 64013f150c2d41965a3d36fa159b55fba52b727a</li>
<li>XGL: dev 4d91a4045078cf6884ee05007eac9eec2b881dbb</li>
<li>LLVM: amd-vulkan-dev 4305f776f1207189667a383e22661ccd1e99cd25</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 900/900 (100.0%)</li>
<li>Failed: 0/900 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 3740/3741 (100.0%)</li>
<li><details><summary>Failed: 1/3741 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.object_management.single.graphics_pipeline
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 3740/3741 (100.0%)</li>
<li><details><summary>Failed: 1/3741 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.object_management.single.graphics_pipeline
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>


</ul></details>

**Comment by JacobHeAMD** [comment]
Hi @trenfouf, looks like this change causes cts failure, could you please have a look?

**Comment by trenouf** [comment]
Hopefully fixed now.

**Comment by amdvlk-admin** [comment]
### Test summary for commit bfae15fe82a7817815b87d20840ca5db9c1adc4e
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev 9a1f99f049f3b417242dc7c09b86c52fd7d9ceaf</li>
<li>SPVGEN: dev 64013f150c2d41965a3d36fa159b55fba52b727a</li>
<li>XGL: dev 4d91a4045078cf6884ee05007eac9eec2b881dbb</li>
<li>LLVM: amd-vulkan-dev 4305f776f1207189667a383e22661ccd1e99cd25</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 900/900 (100.0%)</li>
<li>Failed: 0/900 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 366444/366444 (100.0%)</li>
<li>Failed: 0/366444 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 366444/366444 (100.0%)</li>
<li>Failed: 0/366444 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
Hi, @trenouf , could you move the code in llpcInternal.cpp to llpcUtil.cpp so that

**Comment by JaxLinAMD** [comment]
Hi, @trenouf , could you please move the code in llpcInternal.cpp to llpcUtil.cpp?

**Comment by trenouf** [comment]
> Hi, @trenouf , could you please move the code in llpcInternal.cpp to llpcUtil.cpp?

Done.

**Comment by amdvlk-admin** [comment]
### Test summary for commit bfae15fe82a7817815b87d20840ca5db9c1adc4e
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev 9a1f99f049f3b417242dc7c09b86c52fd7d9ceaf</li>
<li>SPVGEN: dev 64013f150c2d41965a3d36fa159b55fba52b727a</li>
<li>XGL: dev 4d91a4045078cf6884ee05007eac9eec2b881dbb</li>
<li>LLVM: amd-vulkan-dev 4305f776f1207189667a383e22661ccd1e99cd25</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 900/900 (100.0%)</li>
<li>Failed: 0/900 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 366303/366303 (100.0%)</li>
<li>Failed: 0/366303 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 366303/366303 (100.0%)</li>
<li>Failed: 0/366303 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 06a0d65a9de3e2f2a59d286821e49b083946228e
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev 9a1f99f049f3b417242dc7c09b86c52fd7d9ceaf</li>
<li>SPVGEN: dev 64013f150c2d41965a3d36fa159b55fba52b727a</li>
<li>XGL: dev 4d91a4045078cf6884ee05007eac9eec2b881dbb</li>
<li>LLVM: amd-vulkan-dev b9910c8bfcccc63c07c959963fa567120d11f024</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 900/900 (100.0%)</li>
<li>Failed: 0/900 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 294161/294162 (100.0%)</li>
<li><details><summary>Failed: 1/294162 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.image.swapchain_mutable.xlib.2d.r8g8b8a8_unorm_r8g8b8a8_snorm_clear_copy_format_list
Stack trace: Failed to open display: '' at tcuLnxX11.cpp:67
```
</p></details></li>


</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 300669/300670 (100.0%)</li>
<li><details><summary>Failed: 1/300670 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.wsi.xcb.swapchain.render.basic
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>


</ul></details>

**Comment by JacobHeAMD** [comment]
retest this please
